### PR TITLE
REF: collect ops dispatch functions in one place, try to de-duplicate SparseDataFrame methods

### PR DESCRIPTION
**Pull request by jbrockmendel**

### CODE REVIEW

**Comment by jbrockmendel** [comment]
Woops, accidentally pushed some unrelated commits collecting arithmetic tests.

File: pandas/tests/series/test_arithmetic.py [hunk 1]
```diff
@@ -1,18 +1,160 @@
...
+        'ts',
+        [
+            (lambda x: x, lambda x: x * 2, False),
+            (lambda x: x, lambda x: x[::2], False),
+            (lambda x: x, lambda x: 5, True),
+            (lambda x: tm.makeFloatSeries(),
+             lambda x: tm.makeFloatSeries(),
+             True)
+        ])
+    @pytest.mark.parametrize('opname', ['add', 'sub', 'mul', 'floordiv',
```
Comment (jreback) [inline pandas/tests/series/test_arithmetic.py]: ideally switch to our operators fixture

**Review by jreback — CHANGES_REQUESTED** [review]
looks good. just a couple of questions / comments.

File: pandas/core/ops.py [hunk 2]
```diff
@@ -936,6 +936,147 @@ def should_series_dispatch(left, right, op):
...
+    str_rep : str or None, default None
+    axis : {None, 0, 1, "index", "columns"}
+
+    Returns
+    -------
+    DataFrame
+    """
+    # Note: we use iloc to access columns for compat with cases
+    #       with non-unique columns.
+    import pandas.core.computation.expressions as expressions
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/ops.py]: I'm not 100% sure, but I think this is a run-time import to make `import pandas as pd` faster

File: pandas/core/sparse/frame.py [hunk 3]
```diff
@@ -616,24 +596,56 @@ def _combine_match_columns(self, other, func, level=None, try_cast=True):
...
+        if isinstance(other, DataFrame):
+            # i.e. called from _combine_frame
+
+            other_default = getattr(other, 'default_fill_value', np.nan)
+
+            # if the fill values are the same use them? or use a valid one
+            if own_default == other_default:
+                # TOOD: won't this evaluate as False if both are np.nan?
+                fill_value = own_default
+            elif np.isnan(own_default) and not np.isnan(other_default):
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/sparse/frame.py]: At first I thought so, but the module-level docstring says only float64 is supported, so I kept the behavior as-is.  I think the overall takeaway is that this isn't especially well-maintained, and we should all look forward to Sparse EA.

File: pandas/tests/frame/test_arithmetic.py [hunk 4]
```diff
@@ -16,28 +18,86 @@
...
+    def test_timestamp_compare(self):
+        # make sure we can compare Timestamps on the right AND left hand side
+        # GH#4982
+        df = pd. DataFrame({'dates1': pd.date_range('20010101', periods=10),
+                            'dates2': pd.date_range('20010102', periods=10),
+                            'intcol': np.random.randint(1000000000, size=10),
+                            'floatcol': np.random.randn(10),
+                            'stringcol': list(tm.rands(10))})
+        df.loc[np.random.rand(len(df)) > 0.5, 'dates2'] = pd.NaT
+        ops = {'gt': 'lt', 'lt': 'gt', 'ge': 'le', 'le': 'ge', 'eq': 'eq',
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/tests/frame/test_arithmetic.py]: Yah, the point of collecting these arithmetic tests is to parametrize/fixturize and especially de-duplicate them in an upcoming pass.

File: pandas/tests/frame/test_arithmetic.py [hunk 5]
```diff
@@ -16,28 +18,86 @@
...
+    def test_comparison_invalid(self):
+
+        def check(df, df2):
+
+            for (x, y) in [(df, df2), (df2, df)]:
+                # we expect the result to match Series comparisons for
+                # == and !=, inequalities should raise
+                result = x == y
+                expected = pd.DataFrame({col: x[col] == y[col]
+                                         for col in x.columns},
```
Comment (jreback) [inline pandas/tests/frame/test_arithmetic.py]: can you parameterize this (next pass ok)

**Comment by jbrockmendel** [comment]
If it will help, I can separate out the unrelated test parts of this.  There is a bunch of test cleanup to do, and already a healthy number of test-touching PRs in play.

File: pandas/core/sparse/frame.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jreback) [reply pandas/core/sparse/frame.py]: can this be amendednow that that Sparse EA is here? (followup ok too)

**Review by jreback — APPROVED** [review]
comment, but can be a followup, ping on green.

**Comment by jbrockmendel** [comment]
> can this be amendednow that that Sparse EA is here? (followup ok too)

@TomAugspurger any idea about this?  I have no clue

**Comment by jreback** [comment]
also happy to merge this and followup later on sparse refactorings (prob better).

**Comment by jbrockmendel** [comment]
Ping

File: pandas/core/sparse/frame.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/sparse/frame.py]: Sounds good, thanks.

**Comment by jreback** [comment]
can you rebase. the isort is playing havoc :>

**Comment by jreback** [comment]
thanks @jbrockmendel nice as always!

**Pull request merged**